File: catalogador/engine/trd.py
"""TRD/PCD logic (Python port)"""

from typing import Optional, Dict, Any, List
import unicodedata
import json
from pathlib import Path

# Try to load TRD data from engine/data/trd.json if present. Keep a small
# fallback embedded table in case the JSON is missing (backward compatible).
_DATA_FILE = Path(__file__).resolve().parent / "data" / "trd.json"
_LOADED = {}
if _DATA_FILE.exists():
    try:
        with _DATA_FILE.open("r", encoding="utf-8") as fh:
            _LOADED = json.load(fh)
    except Exception:
        _LOADED = {}


def reload_trd_data() -> dict:
    """Reload TRD data from the JSON file and update module-level structures.

    Returns a small summary dict with counts and errors (if any).
    """
    global _LOADED, TRD_TABLA, INVENTARIO_DESCRIPCION, INDICE_TOKENS
    summary = {"loaded": False, "errors": None, "counts": {}}
    try:
        if _DATA_FILE.exists():
            with _DATA_FILE.open("r", encoding="utf-8") as fh:
                _LOADED = json.load(fh)
        else:
            _LOADED = {}

        TRD_TABLA = _get_list("TRD_TABLA", FALLBACK_TRD_TABLA)
        INVENTARIO_DESCRIPCION = _get_dict("INVENTARIO_DESCRIPCION", {})
        INDICE_TOKENS = _get_list("INDICE_TOKENS", [])

        # Basic schema validation for TRD_TABLA entries
        errors = []
        if isinstance(TRD_TABLA, list):
            for i, entry in enumerate(TRD_TABLA):
                if not isinstance(entry, dict):
                    errors.append(f"entry[{i}] is not an object")
                    continue
                if "code" not in entry or not isinstance(entry.get("code"), str):
                    errors.append(f"entry[{i}].code missing or not str")
                if "titulo" not in entry or not isinstance(entry.get("titulo"), str):
                    errors.append(f"entry[{i}].titulo missing or not str")
                if "valor" not in entry or not isinstance(entry.get("valor"), str):
                    errors.append(f"entry[{i}].valor missing or not str")

        summary["loaded"] = True if not errors else False
        summary["counts"] = {
            "trd_tabla": len(TRD_TABLA) if isinstance(TRD_TABLA, list) else 0,
            "indice_tokens": len(INDICE_TOKENS)
            if isinstance(INDICE_TOKENS, list)
            else 0,
            "inventario_descripcion": len(INVENTARIO_DESCRIPCION)
            if isinstance(INVENTARIO_DESCRIPCION, dict)
            else 0,
        }
        if errors:
            summary["errors"] = "; ".join(errors)
    except Exception as e:
        summary["errors"] = str(e)
    return summary


def _get_list(name: str, default: List[Dict[str, Any]]):
    v = _LOADED.get(name)
    return v if isinstance(v, list) else default


def _get_dict(name: str, default: Dict[str, Any]):
    v = _LOADED.get(name)
    return v if isinstance(v, dict) else default


# Fallbacks (small, safe defaults)
FALLBACK_TRD_TABLA: List[Dict[str, Any]] = [
    {
        "code": "CODI/01",
        "titulo": "ACTAS DE SESIONES",
        "valor": "Permanente",
        "ag": 2,
        "ap": 0,
        "oaa": 28,
        "total": 30,
        "asunto": "CONSEJO DIRECTIVO",
    },
    {
        "code": "CODI/02",
        "titulo": "CORRESPONDENCIA",
        "valor": "Temporal",
        "ag": 2,
        "ap": 0,
        "oaa": 6,
        "total": 8,
        "asunto": "CONSEJO DIRECTIVO",
    },
]

TRD_TABLA = _get_list("TRD_TABLA", FALLBACK_TRD_TABLA)
INVENTARIO_DESCRIPCION = _get_dict("INVENTARIO_DESCRIPCION", {})
INDICE_TOKENS = _get_list("INDICE_TOKENS", [])


def normaliza(s: Optional[str]) -> str:
    if not s:
        return ""
    n = unicodedata.normalize("NFD", s)
    n = "".join(ch for ch in n if unicodedata.category(ch) != "Mn")
    return n.upper()


def texto_indice(
    asunto_unidad: Optional[str], titulo_doc: Optional[str], productor: Optional[str]
) -> str:
    parts = [p for p in (asunto_unidad, titulo_doc, productor) if p]
    return " | ".join(normaliza(p) for p in parts)


def tokens_incluidos(texto: str, tokens: List[str]) -> bool:
    return all(normaliza(t) in texto for t in tokens)


def clasificar_por_tokens(
    asunto_unidad: Optional[str], titulo_doc: Optional[str], productor: Optional[str]
) -> Optional[str]:
    txt = texto_indice(asunto_unidad, titulo_doc, productor)
    for r in INDICE_TOKENS:
        if tokens_incluidos(txt, r.get("tokens", [])):
            codes = r.get("codes") or []
            return codes[0] if codes else None
    return None


def buscar_trd_por_codigo(code: str) -> Optional[Dict[str, Any]]:
    if not code:
        return None
    for x in TRD_TABLA:
        if x.get("code") == code:
            return x
    return None


def buscar_trd_por_asunto_titulo(
    asunto: str, serie_sugerida: Optional[str] = None
) -> Optional[Dict[str, Any]]:
    a = normaliza(asunto or "")
    s = normaliza(serie_sugerida or "")
    for x in TRD_TABLA:
        if normaliza(x.get("asunto", "")) in a and normaliza(x.get("titulo", "")) in s:
            return x
    return None


def aplicar_reglas_trd(input_obj: Dict[str, Optional[str]]) -> Dict[str, Any]:
    por_asunto = None
    if input_obj.get("serie") and input_obj.get("asuntoUnidad"):
        por_asunto = buscar_trd_por_asunto_titulo(
            input_obj.get("asuntoUnidad", ""), input_obj.get("serie", "")
        )
    code_tok = clasificar_por_tokens(
        input_obj.get("asuntoUnidad"),
        input_obj.get("titulo"),
        input_obj.get("productor"),
    )
    entry = por_asunto or (buscar_trd_por_codigo(code_tok) if code_tok else None)

    if entry:
        temporalidad = (
            "Permanente" if entry.get("valor") == "Permanente" else "Temporal"
        )
        destino = (
            "Conservación Permanente" if temporalidad == "Permanente" else "Eliminación"
        )
        if temporalidad == "Temporal" and entry.get("oaa", 0) >= 5:
            destino = "Transferencia al Archivo Central"
        return {
            "code": entry.get("code"),
            "tituloSerie": entry.get("titulo"),
            "plazoConservacionAnios": entry.get("total"),
            "temporalidad": temporalidad,
            "destino": destino,
        }

    return {
        "code": None,
        "tituloSerie": input_obj.get("serie") or "Correspondencia",
        "plazoConservacionAnios": 8,
        "temporalidad": "Temporal",
        "destino": "Eliminación",
    }


if __name__ == "__main__":
    samples = [
        {
            "asuntoUnidad": "CONSEJO DIRECTIVO",
            "titulo": "Actas de sesiones",
            "serie": "Actas",
        },
        {
            "asuntoUnidad": "PRESIDENCIA EJECUTIVA",
            "titulo": "Correspondencia general",
            "serie": "Correspondencia",
        },
    ]
    for s in samples:
        print("Input:", s)
        print("Result:", aplicar_reglas_trd(s))

File: _archive/engine_ia/trd.js
/*
  TRD / PCD knowledge module
  - Exports: aplicarReglasTRD, normaliza, clasificarPorTokens, TRD_TABLA, INVENTARIO_DESCRIPCION
  - Self-test when executed with `node trd.js`
*/
'use strict';

// 1) Tabla General de Retención (ANEXO 02)
const TRD_TABLA = [
  { code: 'CODI/01', titulo: 'ACTAS DE SESIONES', valor: 'Permanente', ag: 2, ap: 0, oaa: 28, total: 30, asunto: 'CONSEJO DIRECTIVO' },
  { code: 'CODI/02', titulo: 'CORRESPONDENCIA',    valor: 'Temporal',   ag: 2, ap: 0, oaa: 6,  total: 8,  asunto: 'CONSEJO DIRECTIVO' },
  { code: 'PREJ/01', titulo: 'RESOLUCIONES',       valor: 'Permanente', ag: 2, ap: 0, oaa: 28, total: 30, asunto: 'PRESIDENCIA EJECUTIVA' },
  { code: 'PREJ/02', titulo: 'PRONUNCIAMIENTOS',   valor: 'Permanente', ag: 2, ap: 0, oaa: 28, total: 30, asunto: 'PRESIDENCIA EJECUTIVA' },
  { code: 'PREJ/03', titulo: 'CORRESPONDENCIA',    valor: 'Temporal',   ag: 2, ap: 0, oaa: 6,  total: 8,  asunto: 'PRESIDENCIA EJECUTIVA' },
  { code: 'COIN/01', titulo: 'RESOLUCIONES JEFATURALES', valor: 'Permanente', ag: 2, ap: 8, oaa: 20, total: 30, asunto: 'CONTROL INSTITUCIONAL' },
  { code: 'COIN/02', titulo: 'ACCIONES DE CONTROL',      valor: 'Temporal',   ag: 2, ap: 8, oaa: 5,  total: 15, asunto: 'CONTROL INSTITUCIONAL' },
  { code: 'COIN/03', titulo: 'ACTIVIDADES DE CONTROL',   valor: 'Temporal',   ag: 2, ap: 8, oaa: 5,  total: 15, asunto: 'CONTROL INSTITUCIONAL' },
  { code: 'COIN/04', titulo: 'CORRESPONDENCIA',          valor: 'Temporal',   ag: 2, ap: 8, oaa: 2,  total: 12, asunto: 'CONTROL INSTITUCIONAL' },
  { code: 'COCA/01', titulo: 'CORRESPONDENCIA',          valor: 'Temporal',   ag: 2, ap: 8, oaa: 2,  total: 12, asunto: 'CONTROL DE CALIDAD' },
  { code: 'AGMC/01', titulo: 'INFORMES DE SOCIEDADES AUDITORAS',          valor: 'Temporal', ag: 2, ap: 8, oaa: 5, total: 15, asunto: 'APOYO A LA GESTION Y MEDIDAS CORRECTIVAS' },
  { code: 'AGMC/02', titulo: 'INFORMES DE SEGUIMIENTOS DE MEDIDAS CORRECTIVAS', valor: 'Temporal', ag: 2, ap: 8, oaa: 5, total: 15, asunto: 'APOYO A LA GESTION Y MEDIDAS CORRECTIVAS' },
  { code: 'AGMC/03', titulo: 'VEEDURIAS',                            valor: 'Temporal', ag: 2, ap: 8, oaa: 5, total: 15, asunto: 'APOYO A LA GESTION Y MEDIDAS CORRECTIVAS' },
  { code: 'AGMC/04', titulo: 'CORRESPONDENCIA',                      valor: 'Temporal', ag: 2, ap: 8, oaa: 2, total: 12, asunto: 'APOYO A LA GESTION Y MEDIDAS CORRECTIVAS' },
  { code: 'APGE/01', titulo: 'CORRESPONDENCIA', valor: 'Temporal', ag: 2, ap: 8, oaa: 2, total: 12, asunto: 'APOYO A LA GESTION' },
  { code: 'SEMC/01', titulo: 'CORRESPONDENCIA', valor: 'Temporal', ag: 2, ap: 8, oaa: 2, total: 12, asunto: 'SEGUIMIENTO DE MEDIDAS CORRECTIVAS' },
  { code: 'COCE/01', titulo: 'EXPEDIENTES DE QUEJAS Y DENUNCIAS', valor: 'Temporal', ag: 2, ap: 8, oaa: 5, total: 15, asunto: 'CONTROL CENTRALIZADO' },
  { code: 'COCE/02', titulo: 'HOJAS DE EVALUACIONES',             valor: 'Temporal', ag: 2, ap: 8, oaa: 5, total: 15, asunto: 'CONTROL CENTRALIZADO' },
  { code: 'COCE/03', titulo: 'PAPELES DE TRABAJO',                 valor: 'Temporal', ag: 2, ap: 8, oaa: 5, total: 15, asunto: 'CONTROL CENTRALIZADO' },
  { code: 'COCE/04', titulo: 'CORRESPONDENCIA',                    valor: 'Temporal', ag: 2, ap: 8, oaa: 2, total: 12, asunto: 'CONTROL CENTRALIZADO' },
  { code: 'COC1/01', titulo: 'CORRESPONDENCIA', valor: 'Temporal', ag: 2, ap: 8, oaa: 2, total: 12, asunto: 'CONTROL CENTRALIZADO I' },
  { code: 'COC2/01', titulo: 'CORRESPONDENCIA', valor: 'Temporal', ag: 2, ap: 8, oaa: 2, total: 12, asunto: 'CONTROL CENTRALIZADO II' },
  { code: 'COD1/01', titulo: 'EXPEDIENTES DE QUEJAS Y DENUNCIAS', valor: 'Temporal', ag: 2, ap: 8, oaa: 5, total: 15, asunto: 'CONTROL DESCENTRALIZADO I' },
  { code: 'COD1/02', titulo: 'HOJAS DE EVALUACIONES',             valor: 'Temporal', ag: 2, ap: 8, oaa: 5, total: 15, asunto: 'CONTROL DESCENTRALIZADO I' },
  { code: 'COD1/03', titulo: 'PAPELES DE TRABAJO',                 valor: 'Temporal', ag: 2, ap: 8, oaa: 5, total: 15, asunto: 'CONTROL DESCENTRALIZADO I' },
  { code: 'COD1/04', titulo: 'CORRESPONDENCIA',                    valor: 'Temporal', ag: 2, ap: 8, oaa: 2, total: 12, asunto: 'CONTROL DESCENTRALIZADO I' }
];

// 2) Inventario – descripciones oficiales por código (ANEXO 01)
const INVENTARIO_DESCRIPCION = {
  'CODI/01': 'Actas y acuerdos del Consejo Directivo (con antecedentes).',
  'CODI/02': 'Correspondencia del Consejo Directivo (cartas, oficios, informes, etc.).',
  'PREJ/01': 'Resoluciones de la Presidencia Ejecutiva (con antecedentes y notificación).',
  'PREJ/02': 'Pronunciamientos sobre bases y observaciones de procesos de selección.',
  'PREJ/03': 'Correspondencia de la Presidencia Ejecutiva.',
  'COIN/01': 'Resoluciones jefaturales del Órgano de Control Institucional.',
  'COIN/02': 'Acciones de control posterior (estados fin., presupuestarios, gestión, etc.).',
  'COIN/03': 'Actividades/acciones preventivas de control (veedurías, visitas, etc.).',
  'COIN/04': 'Correspondencia del OCI.',
  'COCE/01': 'Expedientes de quejas y denuncias contra servidores públicos.',
  'COCE/02': 'Hojas de evaluaciones/informativas para control preventivo.',
  'COCE/03': 'Papeles de trabajo de acciones de control.'
};

// 3) Índice alfabético – tokens → código (ANEXO 03)
const INDICE_TOKENS = [
  { tokens: ['ACTAS','CONSEJO DIRECTIVO'], codes: ['CODI/01'] },
  { tokens: ['CORRESPONDENCIA','CONSEJO DIRECTIVO'], codes: ['CODI/02'] },
  { tokens: ['RESOLUCION','PRESIDENCIA EJECUTIVA'], codes: ['PREJ/01'] },
  { tokens: ['PRONUNCIAMIENTO','PRESIDENCIA EJECUTIVA'], codes: ['PREJ/02'] },
  { tokens: ['CORRESPONDENCIA','PRESIDENCIA EJECUTIVA'], codes: ['PREJ/03'] },
  { tokens: ['ACCIONES DE CONTROL'], codes: ['COIN/02'] },
  { tokens: ['ACTIVIDADES DE CONTROL'], codes: ['COIN/03'] },
  { tokens: ['QUEJAS','DENUNCIAS'], codes: ['COCE/01','COD1/01'] },
  { tokens: ['HOJAS','EVALUACIONES'], codes: ['COCE/02','COD1/02'] },
  { tokens: ['PAPELES','TRABAJO'], codes: ['COCE/03','COD1/03'] },
  { tokens: ['CORRESPONDENCIA','CONTROL INSTITUCIONAL'], codes: ['COIN/04'] }
];

function normaliza(s) {
  try {
    return (s || '').normalize('NFD').replace(/\p{Diacritic}/gu, '').toUpperCase();
  } catch (err) {
    // Older Node may not support \p{Diacritic}; fallback: strip combining marks by range
    return (s || '').normalize('NFD').replace(/[\u0300-\u036f]/g, '').toUpperCase();
  }
}

function textoIndice(asuntoUnidad, tituloDoc, productor) {
  return [asuntoUnidad, tituloDoc, productor].filter(Boolean).map(normaliza).join(' | ');
}

function tokensIncluidos(texto, tokens) {
  return tokens.every(t => texto.includes(normaliza(t)));
}

function clasificarPorTokens(asuntoUnidad, tituloDoc, productor) {
  const txt = textoIndice(asuntoUnidad, tituloDoc, productor);
  for (const r of INDICE_TOKENS) {
    if (tokensIncluidos(txt, r.tokens)) return r.codes[0];
  }
  return undefined;
}

function buscarTRDPorCodigo(code) { return TRD_TABLA.find(x => x.code === code); }

function buscarTRDPorAsuntoTitulo(asunto, serieSugerida) {
  const a = normaliza(asunto || '');
  const s = normaliza(serieSugerida || '');
  return TRD_TABLA.find(x => a.includes(normaliza(x.asunto)) && s.includes(normaliza(x.titulo))) || undefined;
}

// Motor final: a partir de (serie, subserie, tipo, título, productor) asigna código, temporalidad, plazos y destino
function aplicarReglasTRD(input) {
  const porAsunto = (input.serie && input.asuntoUnidad) ? buscarTRDPorAsuntoTitulo(input.asuntoUnidad, input.serie) : undefined;
  const codeTok = clasificarPorTokens(input.asuntoUnidad, input.titulo, input.productor);
  const entry = porAsunto || (codeTok ? buscarTRDPorCodigo(codeTok) : undefined);

  if (entry) {
    const temporalidad = entry.valor === 'Permanente' ? 'Permanente' : 'Temporal';
    let destino = temporalidad === 'Permanente' ? 'Conservación Permanente' : 'Eliminación';
    if (temporalidad === 'Temporal' && entry.oaa >= 5) destino = 'Transferencia al Archivo Central';
    return { code: entry.code, tituloSerie: entry.titulo, plazoConservacionAnios: entry.total, temporalidad, destino };
  }

  return { code: undefined, tituloSerie: input.serie || 'Correspondencia', plazoConservacionAnios: 8, temporalidad: 'Temporal', destino: 'Eliminación' };
}

module.exports = {
  TRD_TABLA,
  INVENTARIO_DESCRIPCION,
  INDICE_TOKENS,
  normaliza,
  clasificarPorTokens,
  aplicarReglasTRD,
  buscarTRDPorCodigo,
  buscarTRDPorAsuntoTitulo
};

// Self-test smoke-run when invoked directly
if (require.main === module) {
  console.log('TRD module self-test — running sample classifications');
  const samples = [
    { asuntoUnidad: 'CONSEJO DIRECTIVO', titulo: 'Actas de sesiones', serie: 'Actas' },
    { asuntoUnidad: 'PRESIDENCIA EJECUTIVA', titulo: 'Correspondencia general', serie: 'Correspondencia' },
    { asuntoUnidad: 'CONTROL INSTITUCIONAL', titulo: 'Acciones de control posterior', serie: 'Acciones' },
    { asuntoUnidad: 'CONTROL CENTRALIZADO', titulo: 'Expedientes de quejas y denuncias', serie: 'Quejas' },
    { asuntoUnidad: 'UNIDAD DESCONOCIDA', titulo: 'Carta de presentación', serie: 'Correspondencia' }
  ];

  for (const s of samples) {
    const out = aplicarReglasTRD(s);
    console.log('\nInput:', s);
    console.log('Result:', out);
    if (out.code) {
      console.log('Descripción inventario:', INVENTARIO_DESCRIPCION[out.code] || '(no disponible)');
    }
  }
}
